src/Expander/TC.hs

```diff
@@ -102,22 +100,27 @@ inst :: Scheme Ty -> [Ty] -> Expand Ty
 inst (Scheme n ty) ts
   | length ts /= fromIntegral n =
     throwError $ InternalError "Mismatch in number of type vars"
-  | otherwise = instNorm ty
+  | otherwise = instTy ty
   where
-    instNorm t = do
+    instTy :: Ty -> Expand Ty
+    instTy t = do
       t' <- normType t
-      Ty <$> inst' (unTy t')
+      Ty <$> instTyF (unTy t')
+
+    instTyF :: TyF Ty -> Expand (TyF Ty)
+    instTyF (TyF ctor tArgs) = do
+      let TyF ctor' tArgsPrefix = instCtor ctor
+      tArgsSuffix <- traverse instTy tArgs
+      pure $ TyF ctor' (tArgsPrefix ++ tArgsSuffix)
 
-    inst' (TFun a b) = TFun <$> instNorm a <*> instNorm b
-    inst' (TMacro a) = TMacro <$> instNorm a
-    inst' (TDatatype dt tArgs) = TDatatype dt <$> traverse instNorm tArgs
-    inst' (TSchemaVar i) = pure . unTy $ ts !! fromIntegral i
-    inst' otherTy = pure otherTy
+    instCtor :: TypeConstructor -> TyF Ty
+    instCtor (TSchemaVar i) = unTy $ ts !! fromIntegral i
+    instCtor ctor           = TyF ctor []
 
 
 specialize :: Scheme Ty -> Expand Ty
 specialize sch@(Scheme n _) = do
-  freshVars <- replicateM (fromIntegral n) $ Ty . TMetaVar <$> freshMeta
+  freshVars <- makeTypeMetas n
```

Comment:
there was already a function, `makeTypeMetas`, which did exactly the same thing.

Reply:
:+1: